Improve .NET-Security section about Antiforgery-Tokens

Compufreak345 opened the pull request:
Mainly update information about .NET Core and AJAX with Antiforgery-Tokens and add some reference links.

This PR covers issue #108.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
+After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token.
 
-After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token. Forms must have the requisite helper as seen here:
+If you are using tag-helpers - what is the default for most web project templates - all forms will automatically send the anti-forgery token.
```
`what it` should be `which is`

Compufreak345 replied:
Fixed.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
 
-After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token. Forms must have the requisite helper as seen here:
+If you are using tag-helpers - what is the default for most web project templates - all forms will automatically send the anti-forgery token.
+You can check if tag-helpers are enabled by checking if your main `_ViewImports.cshtml`- file contains:
```
There's an extra hyphen after the filename

Compufreak345 replied:
I removed it.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
+
+#### Using .Net Core 2.0 or full .NET-Framework with AJAX
+
+You will need to attach the anti-forgery token to Ajax requests.
```
A couple of lines above this you have `AJAX`, but here's it's `Ajax`.  I've seen people use both, but `AJAX` is more widely used in the other cheatsheets, so could you change this one for consistency?

Compufreak345 replied:
Makes sense - done.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
````diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
+})
+```
+
+If you are using the full .NET-Framework you can find some snippets [here](https://docs.microsoft.com/de-de/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks#anti-csrf-and-ajax).
````
Change this to the English version (`en-us` rather than `de-de` in the URL).

Compufreak345 replied:
Fixed.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
 
 More information can be found [here](Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md) for Cross-Site Request Forgery.
 
+More information about Anti-Forgery-Tokens can be found [here](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration) for .NET Core and [here](https://docs.microsoft.com/de-de/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks) for the full .NET-Framework.
```
Again, please update to English

Compufreak345 replied:
Fixed.

Compufreak345 commented:
Thanks for the fast review - I fixed the styling issues, please check if it's good now.

rbsec commented:
@Compufreak345 thanks for the updates - looks good to me from a written perspective. I'll let one of the other guys with more of a .NET background comment on the technical side.

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -563,8 +563,12 @@ e.g Startup.cs in the Configure()
+
 DO: Send the anti-forgery token with every POST/PUT request:
 
+#### Using the full .NET-Framework :
```
```suggestion
#### Using .NET Framework:
```

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
+After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token.
 
-After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token. Forms must have the requisite helper as seen here:
+If you are using tag-helpers - which is the default for most web project templates - all forms will automatically send the anti-forgery token.
```
Kindly link tag-helpers to the following [doc](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro)

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
````diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
+@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
+```
+
+`IHtmlHelper.BeginForm` also sends anti-forgery-tokens automatically.
````
Can you add a link to where that is mentioned, so that the reader is able to properly understand what's going underneath if they need to?

Compufreak345 replied:
I added the link where it is mentioned to the beginning of the parent section and this subsection, as it describes TagHelpers as well as the BeginForm-topic.

OK @ThunderSon ?
https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration
`Similarly, IHtmlHelper.BeginForm generates antiforgery tokens by default if the form's method isn't GET.`

ThunderSon replied:
Yes, makes sense. Sadly, the link to `IHtmlGelper.BeginForm` is broken in their docs

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
-And then add the `[AutoValidateAntiforgeryToken]` attribute to the action result.
+And then add the `[AutoValidateAntiforgeryToken]` attribute to the action method. This will validate the token for all HTTP methods other than GET, HEAD, OPTIONS and TRACE.
+
+#### Using .Net Core 2.0 or full .NET-Framework with AJAX
```
```suggestion
#### Using .Net Core 2.0 or .NET Framework with AJAX
```

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
 
 More information can be found [here](Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md) for Cross-Site Request Forgery.
 
+More information about Anti-Forgery-Tokens can be found [here](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration) for .NET Core and [here](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks) for the full .NET-Framework.
```
Mention it as an inline url on [line 566](https://github.com/OWASP/CheatSheetSeries/pull/212/files#diff-00bdcd96a1637f6a84a020c2a6365b9fR566) and remove this line

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
````diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
+})
+```
+
+If you are using the full .NET-Framework you can find some snippets [here](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks#anti-csrf-and-ajax).
````
```suggestion
If you are using the .NET Framework, you can find some code snippets [here](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks#anti-csrf-and-ajax).
```

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,20 +613,55 @@ public void RemoveAntiForgeryCookie(Controller controller)
+
+You will need to attach the anti-forgery token to AJAX requests.
+
+If you are using jQuery in an ASP .NET Core MVC view this can be achieved using this snippet:
```
ASP.NET ; attach it so it becomes `ASP.NET` instead of `ASP .NET`

ThunderSon commented:
Thank you for the updates. There's one outstanding comment. When the code review is done I'll let you know.

Compufreak345 commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
````diff
@@ -609,17 +613,50 @@ public void RemoveAntiForgeryCookie(Controller controller)
 ```
 
-And then add the `[AutoValidateAntiforgeryToken]` attribute to the action result.
+And then add the `[AutoValidateAntiforgeryToken]` attribute to the action method. This will validate the token for all HTTP methods other than GET, HEAD, OPTIONS and TRACE.
````
Good point - I reworked the section to reflect this. What do you think @ThunderSon ?

ThunderSon commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
````diff
@@ -609,17 +613,63 @@ public void RemoveAntiForgeryCookie(Controller controller)
 ```
 
-And then add the `[AutoValidateAntiforgeryToken]` attribute to the action result.
+To automatically validate all requests other than GET, HEAD, OPTIONS and TRACE you need to add a global action filter with the [AutoValidateAntiforgeryToken](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.autovalidateantiforgerytokenattribute?view=aspnetcore-2.2) attribute inside your `Startup.cs` ([Source](https://andrewlock.net/automatically-validating-anti-forgery-tokens- ...
````
```suggestion
To automatically validate all requests other than GET, HEAD, OPTIONS and TRACE you need to add a global action filter with the [AutoValidateAntiforgeryToken](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.autovalidateantiforgerytokenattribute?view=aspnetcore-2.2) attribute inside your `Startup.cs` as mentioned in the following [article](https://andrewlock.net/automatically-validating-anti-forgery-tokens-in-asp-net-core-with-the-autovalidateantiforgerytokenattribute/):
```

jsoref commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,17 +613,63 @@ public void RemoveAntiForgeryCookie(Controller controller)
-NB: You will need to attach the anti-forgery token to Ajax requests.
+#### Using .NET Core 2.0 or later:
+
+After .NET Core 2.0 it is possible to [automatically generate and verify the antiforgery token](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration).
```
`After` here is ambiguous. You might mean `Starting with`...

Compufreak345 replied:
Thanks for the feedback, I learned something as a non-native speaker ;) I changed it.

jsoref commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,17 +613,63 @@ public void RemoveAntiForgeryCookie(Controller controller)
+
+After .NET Core 2.0 it is possible to [automatically generate and verify the antiforgery token](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration).
+
+If you are using [tag-helpers](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro) - which is the default for most web project templates - all forms will automatically send the anti-forgery token.
```
I'd probably use `,`s instead of `-`s. And I'd probably include a `then` before `all`...

Compufreak345 replied:
Done.

jsoref commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,17 +613,63 @@ public void RemoveAntiForgeryCookie(Controller controller)
+`IHtmlHelper.BeginForm` also sends anti-forgery-tokens automatically.
 
-After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token. Forms must have the requisite helper as seen here:
+If you are not using tag-helpers or `IHtmlHelper.BeginForm`, forms must have the requisite helper as seen here:
```
I'd probably include a `then` before `forms`.

Personally, I'm not a fan of `not using ... or`. I'd write something like `if you are using neither ... nor ...`, but people don't really like that. The alternative is `Unless you are using X or Y, you must ...` which is probably better.

Compufreak345 replied:
Done.

Compufreak345 commented:
@jsoref Thanks for the suggestions, I think I implemented all of them.

@ThunderSon Thanks for keeping reviewing ;) I added samples for the attributes.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -563,8 +563,12 @@ e.g Startup.cs in the Configure()
 
 ### Cross-site request forgery
 
+DO NOT: Send sensitive data without validating Anti-Forgery-Tokens ([.NET](https://docs.microsoft.com/en-us/aspnet/web-api/overview/security/preventing-cross-site-request-forgery-csrf-attacks) / [.NET Core](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration)).
```
Better to link to the published MS documentation rather than the GitHub:

https://docs.microsoft.com/en-us/aspnet/core/security/anti-request-forgery?view=aspnetcore-3.0#aspnet-core-antiforgery-configuration

Compufreak345 replied:
Fixed.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,17 +613,96 @@ public void RemoveAntiForgeryCookie(Controller controller)
-NB: You will need to attach the anti-forgery token to Ajax requests.
+#### Using .NET Core 2.0 or later:
+
+Starting with .NET Core 2.0 it is possible to [automatically generate and verify the antiforgery token](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/security/anti-request-forgery.md#aspnet-core-antiforgery-configuration).
```
As above, link to published documents

Compufreak345 replied:
Fixed.

rbsec commented on cheatsheets/DotNet_Security_Cheat_Sheet.md:
```diff
@@ -609,17 +613,96 @@ public void RemoveAntiForgeryCookie(Controller controller)
 
-After .NET Core 2.0 it is possible to automatically generate and verify the antiforgery token. Forms must have the requisite helper as seen here:
+If you are using [tag-helpers](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro), which is the default for most web project templates, then all forms will automatically send the anti-forgery token.
+You can check if tag-helpers are enabled by checking if your main `_ViewImports.cshtml` file contains:
```
Single newlines like this get ignored by markdown, so can be removed. This and the line above just show as one sentence.

Compufreak345 replied:
Fixed.

Compufreak345 commented:
Thanks for the feedback - I think I implemented all requested changes.

ThunderSon commented:
Amazing job 😄
